src/app/data-model/Nullable.h

```diff
@@ -96,22 +98,39 @@ struct Nullable : protected Optional<T>
     // The only fabric-scoped objects in the spec are commands, events and structs inside lists, and none of those can be nullable.
     static constexpr bool kIsFabricScoped = false;
 
-    bool operator==(const Nullable & other) const { return Optional<T>::operator==(other); }
-    bool operator!=(const Nullable & other) const { return Optional<T>::operator!=(other); }
-    bool operator==(const T & other) const { return Optional<T>::operator==(other); }
-    bool operator!=(const T & other) const { return Optional<T>::operator!=(other); }
+    bool operator==(const Nullable<T> & other) const
+    {
+        if (!std::optional<T>::has_value())
+        {
+            return !other.has_value();
+        }
+        return other.has_value() && (*other == **this);
+    }
+    bool operator!=(const Nullable<T> & other) const { return !(*this == other); }
+    bool operator==(const T & other) const { return std::optional<T>::has_value() && (**this == other); }
+    bool operator!=(const T & other) const { return !(*this == other); }
+
+    // backwards compatibility with old chip::Nullable method names
+    //
+    // NOTE: as we transition to std::optional, these should be removed
+    // We expect only `value` and `value_or` to remain as standard names
```

Comment:
Again, I disagree with that expectation.  Why is that the expectation?

Reply:
We seem to have taken `Value` and `ValueOr` as being useful to be taken from chip::Optional. If we are to convert chip::Optional to std::optional it seems to make sense to say "we want to preserve value and value_or" as useful from std::optional.

I am thinking:
   - value and valueor (regardless of syntax) are useful and should exist
   - a `using std::optional<T>::value` seems to likely bring in a more complete declaration of value/value_or than us trying to expose them with a different name
   
 So I have to do extra code to try to expose `Value` instead of `value` and I do not see the benefit: why is it important to name it `Value/ValueOr` instead of `value/value_or`? I agree on "IsNull" is clearer than a "has_value/HasValue" ... however I do not understand the value/valueor bit.
 
 Could you explain your line of thinking?

Reply:
I guess TLDR is I prefer:

```cpp
using std::optional<T>::value_or;
```

over

```cpp
    template <typename... Args>
    auto ValueOr(Args &&... args) -> decltype(std::optional<T>::value_or(std::forward<Args>(args)...))
    {
        return std::optional<T>::value_or(std::forward<Args>(args)...);
    }

    template <typename... Args>
    auto ValueOr(Args &&... args) const -> decltype(std::optional<T>::value_or(std::forward<Args>(args)...))
    {
        return std::optional<T>::value_or(std::forward<Args>(args)...);
    }
```

maybe it can be simplified, however it is ugly and even then I am not sure I covered all the cases transparrently. the using is much simpler.

Reply:
It can definitely be simplified, in that the whole decltype thing is not needed due to type deduction for functions with auto return, and it  covers all cases transparently due to perfect forwarding (though you have to match the actual signatures of value_or, which does not have a non-const version for obvious reasons, but does have a move version which we may or may not care about).  That is you can do:

```
    template <typename... Args>
    constexpr auto ValueOr(Args &&... args) const
    {
        return std::optional<T>::value_or(std::forward<Args>(args)...);
    }
```

and that's all you need for the use cases we have in our tree right now.  Similar for `Value()`.

That said, there is an interesting and subtle difference between `value_or` for `std::optional` and `value_or` for `chip::Optional` and `Nullable` which I just noticed when I was looking into this: `std::optional`'s `value_or` returns by value (makes a copy), whereas `chip::Optional`'s `ValueOr` returns by reference.  Maybe that's an OK change in practice, but we need to watch out for it carefully for non-primitive types, where we might start copying a bunch of stuff we didn't use to.

Back to the "why do we want this?" question: it's a matter of API consistency, basically.  Why `value()` but `IsNull()`?  The inconsistency just looks odd, and to me ironing out the oddness is worth the admittedly Perl-like line of perfect forwarding variadic template.

Since for now we need the aliases anyway, here is my proposal: let's not add the public `value` and `value_or` for now, not add comments about how the existing bits should be removed (again, for now), and not block this PR on this discussion, which we can continue as desired.

Reply:
Also, importantly: I _really_ want the fact that we are using `std::optional` here to be an implementation detail that does not leak through the API, if possible...

Reply:
Updated to only expose Value/ValueOr as the public API and not value/value_or.

I kept the expose of `->` and `*` as unchecked value versions as those seem to be capable to provide slight flash and readability benefits since most of our code does contain `IsNull` guards.

Reply:
Had to make `Value` use operator* so that we get a reference rather than a copy. This has the sideffect of making Value unchecked (likely less flash, however it could be used unsafely).

Reply:
Was looking at the wrong compiler options ... can use `value` just fine. Updated back to it instead of operator*.